### PR TITLE
B 20899 sit end date adjustments main

### PR DESCRIPTION
**Pull request by cameroncaci**
[Integration PR](https://github.com/transcom/mymove/pull/13438)

## Summary

Adds the following service logic for SIT

- If SIT departure date is later than the SIT authorized end date
  - Existing tests updated to support this request (Previously SIT departure was requested without ever even having SIT entry. This broke a lot of things)
- If the SIT departure date is later than the SIT entry date but before the authorized end date
  - Then the SIT authorized end date will be updated to be equal to the departure date
  - This update will run on all service item updates
- If the SIT departure date is earlier than the SIT entry date
  - Then the request will be blocked via a validation error

### First scenario

Given the Prime user is submitting updates to SIT
When the SIT departure date is earlier than the current SIT entry date,
Then no change is made, and the submitter is alerted that the SIT departure date cannot be earlier than the SIT entry date.

1. Log into prime sim
2. Select any move
3. Create new HHG shipment
4. Login as TOO
5. Approve new shipment
6. Switch back to prime sim
7. Add new origin sit service item BUT LEAVE SIT departure Date EMPTY
8. Switch back to TOO
9. Approve said service items associated with origin sit (DOFSIT sets the authorized end date)
10. Switch back to prime sim
11. Edit DOASIT
12. Set departure date to earlier than the SIT start date

### Second scenario

Given the Prime user is submitting updates to SIT
When the SIT departure date is later than the current SIT authorized end date,
Then the SIT authorized end date will not be updated.

1. Log into prime sim
2. Select any move
3. Create new HHG shipment
4. Login as TOO
5. Approve new shipment
6. Switch back to prime sim
7. Add new origin sit service item BUT LEAVE SIT departure Date EMPTY
8. Switch back to TOO
9. Approve said service items associated with origin sit (DOFSIT sets the authorized end date)
10. Switch back to prime sim
11. Edit DOASIT
12. Set departure date to later than the current authorized end date
13. Review the database and see that the authorized end date IS NOT EQUAL TO the SIT departure date

### Third scenario

Given a Prime user is submitting updates to SIT,
When the SIT departure date is earlier than the SIT authorized end date,
Then the SIT authorized end date will recalculate to equal the SIT departure date.

1. Log into prime sim
2. Select any move
3. Create new HHG shipment
4. Login as TOO
5. Approve new shipment
6. Switch back to prime sim
7. Add new origin sit service item BUT LEAVE SIT departure Date EMPTY
8. Switch back to TOO
9. Approve said service items associated with origin sit (DOFSIT sets the authorized end date)
10. Make note of your SIT authorized end date

image

11. Switch back to Prime sim
12. Edit DOASIT
13. Provide a departure date earlier than the authorized end date but after the SIT start date

image

14. Pull your shipment via db or API and make note that the origin_sit_auth_end_date will be equal to the departure date
15. Note that the UI will NOT display your fields properly at this time

image

## Screenshots

image

### CODE REVIEW

File: pkg/services/mto_service_item/mto_service_item_updater.go [hunk 1]
```diff
@@ -332,8 +333,17 @@ func (p *mtoServiceItemUpdater) UpdateMTOServiceItemPrime(
...
+			if err != nil {
+				return updatedServiceItem, err
+			}
 		}
+
+		// If the service item was updated, then it will exist and be passed to this function
+		// We want to chick if the DepartureDate exists, and if it does and it is before
+		// the authorized end date, we need to update the shipment authorized end date
+		// to be equal to the departure date
+		err = setShipmentAuthorizedEndDateToDepartureDate(appCtx, *updatedServiceItem, shipment)
```
Comment (danieljordan-caci) [inline pkg/services/mto_service_item/mto_service_item_updater.go:346]: How come you don't have the `nil` check here like you had in INT?

Screenshot 2024-08-29 at 8 45 21 AM

File: pkg/services/mto_service_item/mto_service_item_validators.go [hunk 2]
```diff
@@ -117,6 +124,14 @@ func (v *primeUpdateMTOServiceItemValidator) validate(appCtx appcontext.AppConte
...
 	}
 
+	// Checks that the SITDepartureDate
+	// - is not later than the authorized end date
+	// - is not before the current entry date
+	err = serviceItemData.checkSITDepartureDate(appCtx)
+	if err != nil {
+		return err
+	}
+
```
Comment (danieljordan-caci) [inline pkg/services/mto_service_item/mto_service_item_validators.go:134]: You have extra comments here compared to INT
Screenshot 2024-08-29 at 8 47 28 AM

File: pkg/services/mto_service_item/mto_service_item_validators.go [hunk 3]
```diff
@@ -379,7 +394,6 @@ func (v *updateMTOServiceItemData) checkNonPrimeFields(_ appcontext.AppContext)
 // checkSITDeparture checks that the service item is a DDDSIT or DOPSIT if the user is trying to update the
 // SITDepartureDate
 func (v *updateMTOServiceItemData) checkSITDeparture(_ appcontext.AppContext) error {
-
```
Comment (danieljordan-caci) [inline pkg/services/mto_service_item/mto_service_item_validators.go:382]: This change is not present in INT
Screenshot 2024-08-29 at 8 48 49 AM

File: pkg/services/mto_service_item/mto_service_item_validators.go [hunk 4]
```diff
@@ -419,6 +433,36 @@ func (v *updateMTOServiceItemData) checkSITEntryDateAndFADD(_ appcontext.AppCont
 	return nil
 }
 
+// checkSITDepartureDate checks that the SITDepartureDate:
+// - is not later than the authorized end date
+// - is not before the current entry date
+func (v *updateMTOServiceItemData) checkSITDepartureDate(_ appcontext.AppContext) error {
```
Comment (danieljordan-caci) [inline pkg/services/mto_service_item/mto_service_item_validators.go:439]: Extra comment line here compared to INT
Screenshot 2024-08-29 at 8 49 25 AM

**Review by danieljordan-caci — CHANGES_REQUESTED** [review]
Some weird little discrepancies just need to be tidied up

File: pkg/services/mto_service_item/mto_service_item_validators.go (hunk 2 above)
Comment (cameroncaci) [reply pkg/services/mto_service_item/mto_service_item_validators.go]: Resulted from merge conflict resolution with integrationTesting. Added a new commit `e11ea6ee68a3f8d4237cc57358f14a4540d27607` to remove this comment in main. This commit will not be present in the intTesting branch, but the code change will be

File: pkg/services/mto_service_item/mto_service_item_updater.go (hunk 1 above)
Comment (cameroncaci) [reply pkg/services/mto_service_item/mto_service_item_updater.go]: [Merge conflict resolution with int testing](https://github.com/transcom/mymove/commit/54a5cb5f62bf71a5b76913b5227b33b4364e9819#diff-175fecaa25f43af40d088729208eb66a692264d4c9cfe7c3fd11d0451ac6cf04)

The current code still has the nil check, but the code organization is what's different. Manually added this back in via commit `a01d22f53334e38852cef613b6e484a3eae6bf21`

File: pkg/services/mto_service_item/mto_service_item_validators.go (hunk 4 above)
Comment (cameroncaci) [reply pkg/services/mto_service_item/mto_service_item_validators.go]: Same as the last one, comment cleanup performed in a merge conflict resolution. Manually added back via `3f3c97f59e562f595bccd0485aa0883cee840be1`

File: pkg/services/mto_service_item/mto_service_item_updater.go [hunk 5]
```diff
@@ -363,6 +365,17 @@ func (p *mtoServiceItemUpdater) UpdateMTOServiceItemPrime(
 		}
 	}
 
+	if updatedServiceItem != nil {
+		// If the service item was updated, then it will exist and be passed to this function
+		// We want to chick if the DepartureDate exists, and if it does and it is before
+		// the authorized end date, we need to update the shipment authorized end date
+		// to be equal to the departure date
+		err = setShipmentAuthorizedEndDateToDepartureDate(appCtx, *updatedServiceItem, shipment)
+		if err != nil {
```
Comment (traskowskycaci) [inline pkg/services/mto_service_item/mto_service_item_updater.go:374]: err check here on line 374 isn't in int

**Review by traskowskycaci — CHANGES_REQUESTED** [review]
One small thing

File: pkg/services/mto_service_item/mto_service_item_updater.go (hunk 5 above)
Comment (cameroncaci) [reply pkg/services/mto_service_item/mto_service_item_updater.go]: Reintroduced in commit `0d64d0571e767968d846e48e700406f07980d154`

**Review by traskowskycaci — APPROVED** [review]
Looks like it matches, other than a few diffs in spacing!

**Review by danieljordan-caci — APPROVED** [review]

**Review by taeJungCaci — APPROVED** [review]
Same as INT

**Pull request merged**